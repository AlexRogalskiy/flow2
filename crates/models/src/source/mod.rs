use crate::tables;

mod loader;
pub mod scenarios;
mod scope;
pub mod specs;
pub mod wrappers;

pub use loader::{FetchResult, LoadError, Loader};
pub use scope::Scope;

pub use specs::*;
pub use wrappers::*;

#[derive(Default, Debug)]
pub struct Tables {
    pub captures: tables::Captures,
    pub collections: tables::Collections,
    pub derivations: tables::Derivations,
    pub endpoints: tables::Endpoints,
    pub errors: tables::Errors,
    pub fetches: tables::Fetches,
    pub imports: tables::Imports,
    pub materializations: tables::Materializations,
    pub nodejs_dependencies: tables::NodeJSDependencies,
    pub projections: tables::Projections,
    pub resources: tables::Resources,
    pub schema_docs: tables::SchemaDocs,
    pub test_steps: tables::TestSteps,
    pub transforms: tables::Transforms,
}

impl Tables {
    // Access Tables as an array of dynamic TableObj instances.
    pub fn as_tables(&self) -> Vec<&dyn tables::TableObj> {
        // This de-structure ensures we can't fail to update if fields change.
        let Tables {
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        } = self;

        vec![
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        ]
    }

    // Access Tables as an array of mutable dynamic TableObj instances.
    pub fn as_tables_mut(&mut self) -> Vec<&mut dyn tables::TableObj> {
        let Tables {
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        } = self;

        vec![
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        ]
    }
}
