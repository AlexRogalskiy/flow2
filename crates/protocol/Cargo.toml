[package]
name = "protocol"
version = "0.0.0"
authors = ["Estuary Technologies, Inc"]
edition = "2018"

[dependencies]
prost = "*"
prost-types = "*"
serde = "*"
tonic = "*"

[dev-dependencies]
insta = "*"
serde_json = "*"

[build-dependencies]
tonic-build = "*"
