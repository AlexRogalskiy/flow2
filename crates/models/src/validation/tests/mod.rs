use crate::source;
use crate::tables;
use crate::validation;

use futures::{future::LocalBoxFuture, FutureExt};
use lazy_static::lazy_static;
use protocol::materialize;
use serde_json::Value;
use std::collections::HashMap;

lazy_static! {
    static ref GOLDEN: Value = serde_yaml::from_slice(include_bytes!("model.yaml")).unwrap();
}

#[test]
fn test_golden_all_visits() {
    let (source_tables, built_tables) = run_test(GOLDEN.clone());
    insta::assert_debug_snapshot!(built_tables);
    insta::assert_debug_snapshot!(source_tables);
}

#[test]
fn test_database_round_trip() {
    let (source_tables, built_tables) = run_test(GOLDEN.clone());

    // Round-trip source and built tables through the database, verifying equality.
    let db = rusqlite::Connection::open(":memory:").unwrap();
    tables::persist_tables(&db, &source_tables.as_tables()).unwrap();
    let mut restored_tables = source::Tables::default();
    tables::load_tables(&db, restored_tables.as_tables_mut().as_mut_slice()).unwrap();

    assert_eq!(
        format!("{:#?}", source_tables),
        format!("{:#?}", restored_tables)
    );

    let db = rusqlite::Connection::open(":memory:").unwrap();
    tables::persist_tables(&db, &built_tables.as_tables()).unwrap();
    let mut restored_tables = validation::Tables::default();
    tables::load_tables(&db, restored_tables.as_tables_mut().as_mut_slice()).unwrap();

    assert_eq!(
        format!("{:#?}", built_tables),
        format!("{:#?}", restored_tables)
    );
}

#[test]
fn test_golden_error() {
    run_test_errors(&GOLDEN, "{}");
}

#[test]
fn test_invalid_collection_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  collections:
    good: &spec
      schema: test://int-string.schema
      key: [/int]
      store: { name: storeEndpoint }

    "": *spec
    bad name: *spec
    bad!name: *spec

    # We require a sequence of non-empty tokens, separated by exactly one '/'.
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string & others.
    testing: *spec

    # Illegal duplicates under naming collation.
    testing/int-sTRinG: *spec
    testing/Int-Halve: *spec
"#,
    );
}

#[test]
fn test_invalid_partition_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      projections:
        good: &spec
          location: /bit
          partition: true

        "": *spec
        inv alid: *spec
        inv!alid: *spec
        inv/alid: *spec

        # Illegal duplicates under collation.
        INT: /int
        bIt: /bit

        # Attempt to re-map a canonical projection.
        str: /int
"#,
    );
}

#[test]
fn test_invalid_transform_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          good: &spec
            source:
              name: testing/int-string
            publish:
              nodeJS: "something something"

          "": *spec
          inv alid: *spec
          inv!alid: *spec
          inv/alid: *spec

          # Illegal duplicate under collation.
          reVeRsEIntString: *spec
"#,
    );
}

#[test]
fn test_invalid_endpoint_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  import:
    - test://more-endpoints
test://more-endpoints:
  endpoints:
    good: &spec
      s3:
        bucket: a-bucket
        prefix: and-prefix

    "": *spec
    inv alid: *spec
    inv!alid: *spec
    inv/alid: *spec

    # Illegal duplicates under collation.
    StoreEndpoinT: *spec
    CAPtUReEndpoINT: *spec
"#,
    );
}

#[test]
fn test_invalid_capture_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  captures:
    good: &spec
      target:
        name: testing/int-string
      pushAPI: {}

    "": *spec
    bad name: *spec
    bad!name: *spec
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string/pull & push.
    testing: *spec

    # Illegal duplicates under naming collation.
    testing/int-sTRinG/pUll: *spec
    testing/Int-strINg/PuSH: *spec
"#,
    );
}

#[test]
fn test_invalid_materialization_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  materializations:
    good: &spec
      source:
        name: testing/int-string
      endpoint:
        name: materializeEndpoint
        config: { fixture: one }

    "": *spec
    bad name: *spec
    bad!name: *spec
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string.
    testing: *spec

    # Illegal duplicate under naming collation.
    testing/int-sTRinG: *spec
"#,
    );
}

#[test]
fn test_transform_source_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source: { name: testinG/Int-String }
          halveSelf:
            source: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_capture_target_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  captures:
    testing/int-string/pull:
      target: { name: testiNg/int-strinK }
    testing/int-string/push:
      target: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_capture_endpoint_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  captures:
    testing/int-string/pull:
      endpoint: { name: CaptureEndpoit }
    testing/int-string/push:
      endpoint: { name: wildlyOffName }
      pushAPI: null
"#,
    );
}

#[test]
fn test_capture_endpoint_wrong_type() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  endpoints:
    captureEndpoint:
      s3: null
      postgres:
        host: a-host
        user: a-user
        password: a-password
"#,
    );
}

#[test]
fn test_capture_target_is_derivation_and_missing_imports() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  import: null
  captures:
    testing/int-string/pull:
      target: { name: testing/int-reverse }
"#,
    );
}

#[test]
fn test_capture_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  captures:
    testing/int-string/another/pull:
      target:
        name: testing/int-string
      endpoint: { name: captureEndpoint }
"#,
    );
}

#[test]
fn test_collection_store_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      store: { name: sToreEndpoin }

test://int-reverse:
  collections:
    testing/int-reverse:
      store: { name: WildlyOffName }
"#,
    );
}

#[test]
fn test_use_without_import() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  import: [] # Clear.
  collections:
    testing/int-string:
      store: { name: s3WithoutImport }

test://int-reverse:
  import: [] # Clear.
  endpoints:
    s3WithoutImport:
      s3:
        bucket: a-bucket
        prefix: and-prefix
"#,
    );
}

#[test]
fn test_schema_fragment_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      schema: test://int-string.schema#/not/found

test://int-string-materialization:
  materializations:
    testing/int-string: null # Omit downstream errors.

test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source:
              schema: test://int-string-len.schema#/not/found
"#,
    );
}

#[test]
fn test_keyed_location_wrong_type() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string.schema:
  properties:
    int: { type: [number, object] }
"#,
    );
}

#[test]
fn test_unknown_locations() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      key: [/int, /unknown/key]
      projections:
        Unknown: /unknown/projection

test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/len, /int, /unknown/shuffle]
"#,
    );
}

#[test]
fn test_shuffle_key_length_mismatch() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/len]
          halveSelf:
            shuffle:
              key: [/len, /int]
"#,
    );
}

#[test]
fn test_shuffle_key_types_mismatch() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/int, /str]
          halveSelf:
            shuffle:
              key: [/str, /int]
"#,
    );
}

#[test]
fn test_shuffle_key_empty() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            shuffle: {key: []}
"#,
    );
}

#[test]
fn test_partition_selections() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source:
              partitions:
                include:
                  bit: [true, 42, ""]
                  Int: [15, true]
                  Unknown: ["whoops"]
                exclude:
                  bit: [false, "a string"]
                  Int: [false, "", 16]
                  AlsoUnknown: ["whoops"]
"#,
    );
}

#[test]
fn test_redundant_source_schema_and_shuffle() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            source:
              name: testing/int-string
              schema: test://int-string.schema
            shuffle:
              key: [/int]
"#,
    );
}

#[test]
fn test_incompatible_collection_store() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  endpoints:
    storeEndpoint:
      s3: null
      postgres:
        host: a-host
        user: a-user
        password: a-password
"#,
    );
}

#[test]
fn test_must_have_update_or_publish() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            publish: null
            update: null
"#,
    );
}

#[test]
fn test_invalid_initial_register() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        register:
          initial: "should be an integer"
"#,
    );
}

#[test]
fn test_shape_inspections() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-len.schema:
  properties:
    int:
      reduce: { strategy: set }
    str:
      reduce: { strategy: sum }
"#,
    );
}

#[test]
fn test_schema_reference_verification() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-len.schema:
  $ref: test://int-string.schema#/whoops
"#,
    );
}

#[test]
fn test_materialization_source_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      source: { name: testiNg/int-strinK }

test://int-halve-materialization:
  materializations:
    testing/int-halve:
      source: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_materialization_endpoint_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      endpoint: { name: MaterializeEndpoit }

test://int-halve-materialization:
  materializations:
    testing/int-halve:
      endpoint: { name: wildlyOffName }
"#,
    );
}

#[test]
fn test_materialization_field_errors() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve-materialization:
  materializations:
    testing/int-halve:
      fields:
        include:
          int: {} # Include and exclude.
          biT: {} # Unknown.
        exclude:
          - BiTT # Unknown.
          - WildlyOffName # Also unknown.
          - int
"#,
    );
}

#[test]
fn test_materialization_driver_returns_error() {
    run_test_errors(
        &GOLDEN,
        r#"
driver:
  materializations:
    one:
      constraints: null
      error: "A driver error!"
"#,
    );
}

#[test]
fn test_materialization_driver_unknown_constraint() {
    run_test_errors(
        &GOLDEN,
        r#"
driver:
  materializations:
    one:
      constraints:
        str: { type: 99, reason: "whoops" }
"#,
    );
}

#[test]
fn test_materialization_driver_conflicts() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      fields:
        include:
          str: {}
        exclude:
          - bit
          - Int
driver:
  materializations:
    one:
      constraints:
        flow_document: { type: 1, reason: "location required" }
        Int: { type: 1, reason: "location required" }

        int: { type: 5, reason: "field unsatisfiable" }
        str: { type: 4, reason: "field forbidden" }
        bit: { type: 0, reason: "field required" }
        Unknown: { type: 0, reason: "whoops" }
"#,
    );
}

#[test]
fn test_test_step_unknown_collection() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - ingest:
          collection: testinG/Int-strin
          documents: []
      - verify:
          collection: wildly/Off/Name
          documents: []
"#,
    );
}

#[test]
fn test_test_step_ingest_schema_error() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - ingest:
          collection: testing/int-string
          documents:
            - {int: 42, str_whoops: "string A", bit: true}
            - {int: 52, str_whoops: "string B", bit: true}
"#,
    );
}

#[test]
fn test_test_step_verify_key_order() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - verify:
          collection: testing/int-string
          documents: [{int: 52}, {int: 62}, {int: 42}]
"#,
    );
}

#[test]
fn test_test_step_verify_selector() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - verify:
          collection: testing/int-string
          documents: []
          partitions:
            include:
              bit: [true, 42, ""]
              Int: [15, true]
              Unknown: ["whoops"]
            exclude:
              bit: [false, "a string"]
              Int: [false, "", 16]
              AlsoUnknown: ["whoops"]
"#,
    );
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
struct MockDriverCalls {
    materializations: HashMap<String, MockMaterializationValidateCall>,
}

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
struct MockMaterializationValidateCall {
    endpoint: source::EndpointType,
    config: serde_json::Value,
    #[serde(default)]
    constraints: HashMap<String, materialize::Constraint>,
    #[serde(default)]
    error: Option<String>,
}

impl validation::Drivers for MockDriverCalls {
    fn validate_materialization(
        &self,
        endpoint_type: source::EndpointType,
        endpoint_config: serde_json::Value,
        _request: materialize::ValidateRequest,
    ) -> LocalBoxFuture<Result<materialize::ValidateResponse, validation::BoxError>> {
        async move {
            for (_key, call) in &self.materializations {
                if (call.endpoint, &call.config) != (endpoint_type, &endpoint_config) {
                    continue;
                }

                if let Some(err) = &call.error {
                    return Err(err.as_str().into());
                } else {
                    return Ok(materialize::ValidateResponse {
                        constraints: call.constraints.clone(),
                    });
                }
            }
            return Err("driver fixture not found".into());
        }
        .boxed_local()
    }
}

fn run_test(mut fixture: Value) -> (source::Tables, validation::Tables) {
    // Extract out driver mock call fixtures.
    let mock_calls = fixture
        .get_mut("driver")
        .map(|d| d.take())
        .unwrap_or_default();
    let mock_calls: MockDriverCalls = serde_json::from_value(mock_calls).unwrap();

    let source_tables = source::scenarios::evaluate_fixtures(Default::default(), &fixture);
    if !source_tables.errors.is_empty() {
        eprint!("{:?}", &source_tables);
        panic!("unexpected fixture load error");
    }
    let built_tables = validation::validate(&mock_calls, &source_tables);

    (source_tables, built_tables)
}

fn run_test_errors(fixture: &Value, patch: &str) {
    let mut fixture = fixture.clone();
    let patch: Value = serde_yaml::from_str(patch).unwrap();
    json_patch::merge(&mut fixture, &patch);

    let (_, built_tables) = run_test(fixture);
    insta::assert_debug_snapshot!(built_tables.errors);
}
