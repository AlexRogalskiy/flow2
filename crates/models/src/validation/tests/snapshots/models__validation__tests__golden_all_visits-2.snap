---
source: crates/models/src/validation/tests/mod.rs
expression: source_tables
---
Tables {
    captures: [
        Capture {
            scope: test://int-string-capture#/captures/testing~1int-string~1pull,
            capture: testing/int-string/pull,
            collection: testing/int-string,
            allow_push: 0,
            endpoint: captureEndpoint,
            patch_config: {},
        },
        Capture {
            scope: test://int-string-capture#/captures/testing~1int-string~1push,
            capture: testing/int-string/push,
            collection: testing/int-string,
            allow_push: 1,
            endpoint: NULL,
            patch_config: {},
        },
    ],
    collections: [
        Collection {
            scope: test://int-reverse#/collections/testing~1int-reverse,
            collection: testing/int-reverse,
            schema: test://int-string.schema,
            key: ["/int"],
            store_endpoint: storeEndpoint,
            store_patch_config: {},
        },
        Collection {
            scope: test://int-string#/collections/testing~1int-string,
            collection: testing/int-string,
            schema: test://int-string.schema,
            key: ["/int"],
            store_endpoint: storeEndpoint,
            store_patch_config: {},
        },
        Collection {
            scope: test://int-halve#/collections/testing~1int-halve,
            collection: testing/int-halve,
            schema: test://int-string-len.schema,
            key: ["/int"],
            store_endpoint: storeEndpoint,
            store_patch_config: {},
        },
    ],
    derivations: [
        Derivation {
            scope: test://int-reverse#/collections/testing~1int-reverse/derivation,
            derivation: testing/int-reverse,
            register_schema: test://int-reverse?ptr=/collections/testing~1int-reverse/derivation/register/schema,
            register_initial: null,
        },
        Derivation {
            scope: test://int-halve#/collections/testing~1int-halve/derivation,
            derivation: testing/int-halve,
            register_schema: test://int-halve?ptr=/collections/testing~1int-halve/derivation/register/schema,
            register_initial: 42,
        },
    ],
    endpoints: [
        Endpoint {
            scope: test://root#/endpoints/captureEndpoint,
            endpoint: captureEndpoint,
            endpoint_type: "S3",
            base_config: {"bucket":"a-bucket","prefix":"and-prefix"},
        },
        Endpoint {
            scope: test://root#/endpoints/materializeEndpoint,
            endpoint: materializeEndpoint,
            endpoint_type: "S3",
            base_config: {"bucket":"a-bucket","prefix":"and-prefix"},
        },
        Endpoint {
            scope: test://root#/endpoints/storeEndpoint,
            endpoint: storeEndpoint,
            endpoint_type: "S3",
            base_config: {"bucket":"a-bucket","prefix":"and-prefix"},
        },
    ],
    errors: [],
    fetches: [
        Fetch {
            resource: test://root,
        },
        Fetch {
            resource: test://int-string,
        },
        Fetch {
            resource: test://int-reverse,
        },
        Fetch {
            resource: test://int-halve,
        },
        Fetch {
            resource: test://int-string-capture,
        },
        Fetch {
            resource: test://int-string-materialization,
        },
        Fetch {
            resource: test://int-halve-materialization,
        },
        Fetch {
            resource: test://int-string-tests,
        },
        Fetch {
            resource: test://int-string.schema,
        },
        Fetch {
            resource: test://int-string-len.schema,
        },
    ],
    imports: [
        Import {
            scope: test://int-string#/import/0,
            from_resource: test://int-string,
            to_resource: test://int-halve,
        },
        Import {
            scope: test://int-reverse#/import/0,
            from_resource: test://int-reverse,
            to_resource: test://int-string,
        },
        Import {
            scope: test://int-reverse#/collections/testing~1int-reverse/schema,
            from_resource: test://int-reverse,
            to_resource: test://int-string.schema,
        },
        Import {
            scope: test://int-reverse#/collections/testing~1int-reverse/derivation/register/schema,
            from_resource: test://int-reverse,
            to_resource: test://int-reverse?ptr=/collections/testing~1int-reverse/derivation/register/schema,
        },
        Import {
            scope: test://root#/import/1,
            from_resource: test://root,
            to_resource: test://int-reverse,
        },
        Import {
            scope: test://int-halve#/collections/testing~1int-halve/derivation/register/schema,
            from_resource: test://int-halve,
            to_resource: test://int-halve?ptr=/collections/testing~1int-halve/derivation/register/schema,
        },
        Import {
            scope: test://int-halve#/collections/testing~1int-halve/derivation/transform/halveIntString/source/schema,
            from_resource: test://int-halve,
            to_resource: test://int-string-len.schema,
        },
        Import {
            scope: test://int-string-capture#/import/0,
            from_resource: test://int-string-capture,
            to_resource: test://int-string,
        },
        Import {
            scope: test://root#/import/3,
            from_resource: test://root,
            to_resource: test://int-string-capture,
        },
        Import {
            scope: test://int-string-materialization#/import/0,
            from_resource: test://int-string-materialization,
            to_resource: test://int-string,
        },
        Import {
            scope: test://root#/import/4,
            from_resource: test://root,
            to_resource: test://int-string-materialization,
        },
        Import {
            scope: test://int-halve-materialization#/import/0,
            from_resource: test://int-halve-materialization,
            to_resource: test://int-halve,
        },
        Import {
            scope: test://root#/import/5,
            from_resource: test://root,
            to_resource: test://int-halve-materialization,
        },
        Import {
            scope: test://int-string-tests#/import/0,
            from_resource: test://int-string-tests,
            to_resource: test://int-string,
        },
        Import {
            scope: test://root#/import/6,
            from_resource: test://root,
            to_resource: test://int-string-tests,
        },
        Import {
            scope: test://int-string#/collections/testing~1int-string/schema,
            from_resource: test://int-string,
            to_resource: test://int-string.schema,
        },
        Import {
            scope: test://root#/import/0,
            from_resource: test://root,
            to_resource: test://int-string,
        },
        Import {
            scope: test://int-string-len.schema#/$ref,
            from_resource: test://int-string-len.schema,
            to_resource: test://int-string.schema,
        },
        Import {
            scope: test://int-halve#/collections/testing~1int-halve/schema,
            from_resource: test://int-halve,
            to_resource: test://int-string-len.schema,
        },
        Import {
            scope: test://root#/import/2,
            from_resource: test://root,
            to_resource: test://int-halve,
        },
    ],
    materializations: [
        Materialization {
            scope: test://int-string-materialization#/materializations/testing~1int-string,
            materialization: testing/int-string,
            collection: testing/int-string,
            endpoint: materializeEndpoint,
            patch_config: {"fixture":"one"},
            field_selector: {"include":{"str":{"pass":"through"}},"exclude":["bit"],"recommended":true},
        },
        Materialization {
            scope: test://int-halve-materialization#/materializations/testing~1int-halve,
            materialization: testing/int-halve,
            collection: testing/int-halve,
            endpoint: materializeEndpoint,
            patch_config: {"fixture":"two"},
            field_selector: {"include":{"Len":{}},"exclude":[],"recommended":false},
        },
    ],
    nodejs_dependencies: [],
    projections: [
        Projection {
            scope: test://int-string#/collections/testing~1int-string/projections/Int,
            collection: testing/int-string,
            field: Int,
            location: /int,
            partition: 0,
            user_provided: 1,
        },
        Projection {
            scope: test://int-string#/collections/testing~1int-string/projections/bit,
            collection: testing/int-string,
            field: bit,
            location: /bit,
            partition: 1,
            user_provided: 1,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve/projections/Extra,
            collection: testing/int-halve,
            field: Extra,
            location: /extra,
            partition: 0,
            user_provided: 1,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve/projections/Len,
            collection: testing/int-halve,
            field: Len,
            location: /len,
            partition: 1,
            user_provided: 1,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve/projections/Root,
            collection: testing/int-halve,
            field: Root,
            location: ,
            partition: 0,
            user_provided: 1,
        },
    ],
    resources: [
        Resource {
            resource: test://root,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-string,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-reverse,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-reverse?ptr=/collections/testing~1int-reverse/derivation/register/schema,
            content_type: "JsonSchema",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-halve,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-halve?ptr=/collections/testing~1int-halve/derivation/register/schema,
            content_type: "JsonSchema",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-string-capture,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-string-materialization,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-halve-materialization,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-string-tests,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-string.schema,
            content_type: "JsonSchema",
            content: ".. binary ..",
        },
        Resource {
            resource: test://int-string-len.schema,
            content_type: "JsonSchema",
            content: ".. binary ..",
        },
    ],
    schema_docs: [
        SchemaDoc {
            schema: test://int-reverse?ptr=/collections/testing~1int-reverse/derivation/register/schema,
            dom: true,
        },
        SchemaDoc {
            schema: test://int-halve?ptr=/collections/testing~1int-halve/derivation/register/schema,
            dom: {"type":"integer"},
        },
        SchemaDoc {
            schema: test://int-string.schema,
            dom: {"properties":{"bit":{"type":"boolean"},"int":{"type":"integer"},"str":{"type":"string"}},"required":["int","str","bit"],"type":"object"},
        },
        SchemaDoc {
            schema: test://int-string-len.schema,
            dom: {"$ref":"test://int-string.schema","additionalProperties":{"type":"boolean"},"properties":{"bit":{},"int":{},"len":{"type":"integer"},"str":{}},"required":["len"]},
        },
    ],
    test_steps: [
        TestStep {
            scope: test://int-string-tests#/tests/A%20Test/0,
            test: A Test,
            step_index: 0,
            step: {"ingest":{"collection":"testing/int-string","documents":[{"bit":true,"int":42,"str":"string A"},{"bit":true,"int":52,"str":"string B"}]}},
        },
        TestStep {
            scope: test://int-string-tests#/tests/A%20Test/1,
            test: A Test,
            step_index: 1,
            step: {"verify":{"collection":"testing/int-string","documents":[{"str":"string A"},{"str":"string B"}],"partitions":{"include":{"bit":[true]},"exclude":{}}}},
        },
    ],
    transforms: [
        Transform {
            scope: test://int-reverse#/collections/testing~1int-reverse/derivation/transform/reverseIntString,
            transform: reverseIntString,
            derivation: testing/int-reverse,
            source_collection: testing/int-string,
            source_partitions: NULL,
            source_schema: NULL,
            shuffle_key: NULL,
            shuffle_lambda: NULL,
            shuffle_hash: "None",
            read_delay_seconds: NULL,
            priority: 0,
            update_lambda: NULL,
            publish_lambda: {"nodeJS":"return [{\n  int: source.int,\n  str: source.str.reverse(),\n  len: source.str.length,\n}];\n"},
        },
        Transform {
            scope: test://int-halve#/collections/testing~1int-halve/derivation/transform/halveIntString,
            transform: halveIntString,
            derivation: testing/int-halve,
            source_collection: testing/int-string,
            source_partitions: {"include":{"bit":[true]},"exclude":{"bit":[false]}},
            source_schema: test://int-string-len.schema,
            shuffle_key: ["/len","/int"],
            shuffle_lambda: NULL,
            shuffle_hash: "None",
            read_delay_seconds: NULL,
            priority: 0,
            update_lambda: NULL,
            publish_lambda: {"nodeJS":"... transform int-string ...\n"},
        },
        Transform {
            scope: test://int-halve#/collections/testing~1int-halve/derivation/transform/halveSelf,
            transform: halveSelf,
            derivation: testing/int-halve,
            source_collection: testing/int-halve,
            source_partitions: NULL,
            source_schema: NULL,
            shuffle_key: ["/len","/int"],
            shuffle_lambda: NULL,
            shuffle_hash: "None",
            read_delay_seconds: NULL,
            priority: 0,
            update_lambda: NULL,
            publish_lambda: {"nodeJS":"... transform self ...\n"},
        },
    ],
}
